File: app/src/main/java/ar/edu/utn/frba/dadm/clases2018c1/clases_2018c1/MainActivity.java
package ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.AppCompatButton;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.messaging.FirebaseMessaging;

import ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1.preferences.MyPreferences;

public class MainActivity extends AppCompatActivity {
    TextView firebaseTokenTextView;
    EditText topicEditText;
    AppCompatButton reloadButton;
    AppCompatButton copyButton;
    AppCompatButton subscribeButton;
    View topicContainer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        reloadButton = findViewById(R.id.reload_button);
        copyButton = findViewById(R.id.copy_button);
        subscribeButton = findViewById(R.id.subscribe_button);
        topicEditText = findViewById(R.id.topic);
        topicContainer = findViewById(R.id.topic_container);

        setFirebaseTokenInView();

        reloadButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                setFirebaseTokenInView();
            }
        });

        copyButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                copyTokenToClipboard();
            }
        });

        subscribeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                subscribeToTopic();
            }
        });
    }

    private void setFirebaseTokenInView() {
        String firebaseToken = MyPreferences.getFirebaseToken(this);

        if(firebaseToken != null){
            firebaseTokenTextView = findViewById(R.id.firebaseToken);
            firebaseTokenTextView.setText(firebaseToken);
            reloadButton.setVisibility(View.GONE);
            copyButton.setVisibility(View.VISIBLE);
            topicContainer.setVisibility(View.VISIBLE);
        } else {
            copyButton.setVisibility(View.GONE);
            topicContainer.setVisibility(View.GONE);
            reloadButton.setVisibility(View.VISIBLE);
        }
    }

    private void subscribeToTopic() {
        String topic = topicEditText.getText().toString();
        FirebaseMessaging.getInstance().subscribeToTopic(topic);
        Toast.makeText(this, "Subscripto a " + topic, Toast.LENGTH_SHORT).show();
        topicEditText.setText("");
    }

    public void copyTokenToClipboard() {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("Firebase token", firebaseTokenTextView.getText());
        clipboard.setPrimaryClip(clip);
    }
}

File: app/src/main/java/ar/edu/utn/frba/dadm/clases2018c1/clases_2018c1/MainActivity.kt
package ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.*
import ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1.api.Callback
import ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1.api.OmdbApi
import ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1.api.responses.Movie
import ar.edu.utn.frba.dadm.clases2018c1.clases_2018c1.api.responses.MovieSearch
import java.util.concurrent.Executors

class MainActivity : AppCompatActivity() {

    val MOVIES = "MOVIES"

    lateinit var titleInput: EditText
    lateinit var progressBar: ProgressBar
    lateinit var findButton: View
    lateinit var recyclerView: RecyclerView
    lateinit var movies: ArrayList<Movie>
    lateinit var adapter: MovieAdapter

    private val executor = Executors.newSingleThreadExecutor()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        movies = savedInstanceState?.getSerializable(MOVIES) as ArrayList<Movie>? ?: ArrayList()
        setupViews()
    }

    fun setupViews() {
        titleInput = findViewById(R.id.titleInput)
        recyclerView = findViewById(R.id.recyclerView)
        progressBar = findViewById(R.id.progressBar)
        findButton = findViewById(R.id.findButton)
        findButton.setOnClickListener {
            findMovie()
        }
        recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = MovieAdapter(this, movies, false)
        recyclerView.adapter = adapter
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(MOVIES, movies)
    }

    fun findMovie() {
        setLoading(true)
        val apiKey = "7a9f6b43"
        val title = titleInput.text.toString()
        executor.submit(OmdbApi.getMovies(apiKey, title, object : Callback<MovieSearch> {
            override fun onSuccess(movieSearch: MovieSearch) {

                runOnUiThread {
                    setLoading(false)
                    if (movieSearch.response) {
                        for (movie in movieSearch.search!!){
                            addMovie(movie)
                        }
                    }
                    else {
                        showError(null)
                    }
                }
            }
            override fun onError(error: Exception?) {
                runOnUiThread {
                    setLoading(false)
                    showError(error)
                }
            }
        }))
    }

    fun setLoading(isLoading: Boolean) {
        progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
        findButton.isEnabled = !isLoading

    }

    fun addMovie(movie: Movie) {
        movies.add(0, movie)
        adapter.notifyItemInserted(0)
        recyclerView.scrollToPosition(0)
    }

    fun showError(error: Exception?) {
        Toast.makeText(this, error?.toString() ?: "Error", Toast.LENGTH_LONG).show()
    }
}
